flink-runtime/src/main/java/org/apache/flink/runtime/rest/handler/router/RouterHandler.java

```diff
@@ -0,0 +1,109 @@
+import org.apache.flink.shaded.netty4.io.netty.channel.SimpleChannelInboundHandler;
+import org.apache.flink.shaded.netty4.io.netty.handler.codec.http.DefaultFullHttpResponse;
+import org.apache.flink.shaded.netty4.io.netty.handler.codec.http.HttpHeaders;
+import org.apache.flink.shaded.netty4.io.netty.handler.codec.http.HttpMethod;
+import org.apache.flink.shaded.netty4.io.netty.handler.codec.http.HttpRequest;
+import org.apache.flink.shaded.netty4.io.netty.handler.codec.http.HttpResponseStatus;
+import org.apache.flink.shaded.netty4.io.netty.handler.codec.http.HttpVersion;
+import org.apache.flink.shaded.netty4.io.netty.handler.codec.http.QueryStringDecoder;
+
+import java.util.Map;
+
+import static java.util.Objects.requireNonNull;
+
+/**
+ * This class replaces the standard error response to be identical with those sent by the {@link AbstractRestHandler}.
+ */
+public class RouterHandler extends SimpleChannelInboundHandler<HttpRequest> {
+	public static final String ROUTER_HANDLER_NAME = RouterHandler.class.getName() + "_ROUTER_HANDLER";
+	public static final String ROUTED_HANDLER_NAME = RouterHandler.class.getName() + "_ROUTED_HANDLER";
+
+	private final Map<String, String> responseHeaders;
+	private final Router router;
+
+	public RouterHandler(Router router, final Map<String, String> responseHeaders) {
+		this.router = requireNonNull(router);
+		this.responseHeaders = requireNonNull(responseHeaders);
+	}
+
+	public String getName() {
+		return ROUTER_HANDLER_NAME;
+	}
+
+	@Override
+	protected void channelRead0(ChannelHandlerContext channelHandlerContext, HttpRequest httpRequest) throws Exception {
+		if (HttpHeaders.is100ContinueExpected(httpRequest)) {
+			channelHandlerContext.writeAndFlush(new DefaultFullHttpResponse(HttpVersion.HTTP_1_1, HttpResponseStatus.CONTINUE));
+			return;
+		}
+
+		// Route
+		HttpMethod method = httpRequest.getMethod();
+		QueryStringDecoder qsd = new QueryStringDecoder(httpRequest.getUri());
+		RouteResult routeResult = router.route(method, qsd.path(), qsd.parameters());
+
+		if (routeResult == null) {
+			respondNotFound(channelHandlerContext, httpRequest);
+			return;
+		}
+
+		routed(channelHandlerContext, routeResult, httpRequest);
+	}
+
+	private void routed(
+			ChannelHandlerContext channelHandlerContext,
+			RouteResult routeResult,
+			HttpRequest httpRequest) throws Exception {
+		ChannelInboundHandler handler = (ChannelInboundHandler) routeResult.target();
+
+		// The handler may have been added (keep alive)
+		ChannelPipeline pipeline     = channelHandlerContext.pipeline();
+		ChannelHandler addedHandler = pipeline.get(ROUTED_HANDLER_NAME);
+		if (handler != addedHandler) {
+			if (addedHandler == null) {
+				pipeline.addAfter(ROUTER_HANDLER_NAME, ROUTED_HANDLER_NAME, handler);
+			} else {
+				pipeline.replace(addedHandler, ROUTED_HANDLER_NAME, handler);
+			}
+		}
+
+		channelHandlerContext.fireChannelRead(new RoutedRequest(routeResult, httpRequest).retain());
```

Comment:
changed into
```
RoutedRequest<?> request = new RoutedRequest<>(routeResult, httpRequest);
channelHandlerContext.fireChannelRead(request.retain());
```
I like having `.retain()` calls to be in the same line where they are actually designed for, since it's kind of self documenting which line is the reason behind retaining.

Reply:
:+1: